File: 0150-evaluate-reverse-polish-notation/0150-evaluate-reverse-polish-notation.cpp
class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        stack<int> st;

        for (const string& tok : tokens) {
            if (tok == "+" || tok == "-" || tok == "*" || tok == "/") {
                // Pop the top two operands
                int b = st.top(); st.pop();
                int a = st.top(); st.pop();
                int res = 0;

                if (tok == "+")      res = a + b;
                else if (tok == "-") res = a - b;
                else if (tok == "*") res = a * b;
                else if (tok == "/") res = a / b;  // truncates toward zero

                st.push(res);
            } else {
                // It's a number
                st.push(stoi(tok));
            }
        }

        return st.top();
    }
};

File: 0150-evaluate-reverse-polish-notation/0150-evaluate-reverse-polish-notation.py
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []
        ops = {'+', '-', '*', '/'}
        
        for tok in tokens:
            if tok in ops:
                b = stack.pop()
                a = stack.pop()
                if tok == '+':
                    stack.append(a + b)
                elif tok == '-':
                    stack.append(a - b)
                elif tok == '*':
                    stack.append(a * b)
                else:
                    # truncate toward zero
                    stack.append(int(a / b))
            else:
                stack.append(int(tok))
        
        return stack[0]     